### ISSUE TITLE
InputObjectType defaultValue is not getting used

### ISSUE THREAD

**Issue by Jimexist** [graphql-java/graphql-java#152]
Currently in `graphql.execution.ValuesResolver#coerceValueAstForInputObject` the `inputValue`'s fields are iterated, not `type`'s, i.e.:

``` java
   private Object coerceValueAstForInputObject(GraphQLInputObjectType type, ObjectValue inputValue, Map<String, Object> variables) {
        Map<String, Object> result = new LinkedHashMap<String, Object>();

        for (ObjectField objectField : inputValue.getObjectFields()) {
            GraphQLInputObjectField inputObjectField = type.getField(objectField.getName());
            // illegal field ... no corresponding key in the schema
            if (inputObjectField == null) continue;
            Object fieldValue = coerceValueAst(inputObjectField.getType(), objectField.getValue(), variables);
            if (fieldValue == null) {
                fieldValue = inputObjectField.getDefaultValue();
            }
            result.put(objectField.getName(), fieldValue);
        }
        return result;
    }
```

That means, if the a field in `type` is absent in `inputValue`, its default value will not gets populated.

**Comment by trevor-morris** [comment]
Interesting; this is somewhat related to #106 

I think that the handling of input objects is inconsistent at the moment. An argument which is an input object can be given a value either via an object literal or a variable value. In either case the value may be incomplete - that is it may omit fields present in the input object type. At the moment I believe the following is true:

If given via an object literal then any fields missing in the literal are omitted from the argument value. But default values are not applied for the missing field and I suspect that if one of the missing fields is "non null" then that won't get detected either (that's just a guess, based on looking at the code).

If given via a variable value then any fields missing in the variable value are still set in the argument value - either to null, or to the default value, if present. Omitting a non null field is also correctly detected.

I think these two problems are mirror images of each other. In one case (literal -> argument value) we only copy the fields actually present in the input, which means we (correctly) don't copy missing fields but also forget to apply default fields and non null checks for those missing fields. In the other case (variable value -> argument value) we copy all the fields, whether or not they're in the input. That means we correctly apply default values and non null checks, even for missing fields, but we incorrectly set the value of any nullable, non defaulted missing fields to null in the argument value.

I have a pull request in with what I think is a fix for #106; if I get chance in the next day or so I'll alter it to try and fix this issue, #152, as well.

One thing I was unsure about; if the literal or variable value explicitly sets a field to null should the default value apply? At first I thought not. But reading the GraphQL spec it sounds like items with a default values should be thought of as "non null" so I believe that the default should apply even in this case. So, for example, if you have an input object type with a nullable string field "x" which has default value "DefaultX" then specifying { x: null } should result it "x" having the value "DefaultX".

**Comment by dminkovsky** [comment]
Thanks for posting @Jimexist.
